docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,285 @@
+  * For table indices, we replace current `ColumnValue` with `sortKey` without additional change:
+    - For unique indices:
+	```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKey
+      Value: rowID
+	```
+    - For non-unique indices:
+    ```
+      Key: tablePrefix{tableID}_indexPrefixSep{indexID}_sortKeys_rowID
+      Value: null
+	```
+
+Pros:
+    The current encoding layout keeps unchanged.
+
+Cons:
+    Since index value can not be recovered from `sortKey`, extra table lookup is needed for current index-only queries, this may lead to performance regressions and the "covering index" can not work. The extra table lookup also requires corresponding changes to the TiDB optimizer.
+
+The reason to reject this option is, as a distributed system, the cost of potential table lookup high so that it should be avoided it as much as possible. What's more, adding a covering index is a common SQL optimization method for the DBAs: table lookup can not be avoided through covering index is counter-intuitive.
+
+## Compatibility
+
+### Compatibility with MySQL
+
+As stated before, MySQL has various different collation implementations. For example, in MySQL 5.7, the default collation of `utf8mb4` is `utf8mb4_general_ci`, which has two main flaws:
+
+  * It only implement the `simple mappings` of UCA.
+  * It used a very old version of UCA.
+
+In MySQL 8, the default collation of `utf8mb4` has been changed to `utf8mb4_0900_ai_ci`, in which the Unicode version is upgraded to 9.0, with full UCA implementation.
+
+Should TiDB aim to support all collation of MySQL?
+
+  * If TiDB supports the latest collations only(for example, support `utf8mb4_0900_ai_ci` only and treat `utf8mb4_general_ci` as one of its alias). This seems to be 'standard-correct' but may lead to potential compatibility issues.
+  * If TiDB supports all the collations, all of their implementations should be analyzed carefully. That is a lot of work.
+
+### Compatibility between TiDB versions
+
+In this proposal, both of the compatibility issues can be solved in the new version of TiDB, and with the help of the version of metadata in TiDB, we can always indicate if a collation is old(the "fake one") or new(the "real one"). However, here we propose several requirements below and careful considerations are needed to meet them as many as possible:
+
+  1. For an existing TiDB cluster, its behavior remains unchanged after upgrading to newer version
+    - a. After upgrade, the behavior of existing tables remains unchanged.
+    - b. After upgrade, the behavior of newly-created tables remains unchanged.
+    - c. If a TiDB cluster is replicating the data to a MySQL instance through Binlog, all  replications remain unchanged.
+  2. Users of MySQL who requires collations can move to TiDB without modifying their scripts/applications.
+  3. Users of TiDB can distinguish the tables with old collation and new collations easily, if they both exist in a TiDB cluster.
```

Comment:
Are we planning to make it possible to have tables with "old collation" and "new collation" coexist on the same cluster? 

The current implementation reads the collation info on bootstrap from the `mysql.tidb` table, is that value supposed to be mutable?

Reply:
> Are we planning to make it possible to have tables with "old collation" and "new collation" coexist on the same cluster?

Yes, but currently we don't have a detailed roadmap for it.